### PR TITLE
Model Leftovers

### PR DESCRIPTION
**Pull request by afein**
just dropping this here for review while I'm cleaning it up.

\cc @vishh @mvdan 

Please take a look before EOD today.
The tests are not expected to work as of this point, this would involve about ~1.5k LOC changes as well.

### CODE REVIEW

File: api/v1/model_handlers.go [hunk 1]
```diff
@@ -371,7 +449,17 @@ func (a *Api) allFreeContainers(request *restful.Request, response *restful.Resp
 		return
 	}
 	node := request.PathParameter("node-name")
-	response.WriteEntity(cluster.GetFreeContainers(node))
+	response.WriteEntity(makeExternalEntityList(cluster.GetFreeContainers(node)))
+}
+
+// nodePods returns a list of all the available API paths that are available for a node.
+func (a *Api) nodePods(request *restful.Request, response *restful.Response) {
+	cluster := a.manager.GetCluster()
```
Comment (vishh) [inline api/v1/model_handlers.go]: Does this Getter copy the entire model?

File: deploy/kube-config/kubedash/heapster-controller.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (vishh) [inline deploy/kube-config/kubedash/heapster-controller.json]: Why is this required if the model is turned on by default?

File: model/aggregation.go [hunk 3]
```diff
@@ -182,33 +183,44 @@ func (rc *realCluster) aggregateMetrics(target *InfoType, sources []*InfoType) e
...
-	// Put all the new values in the TimeStores under target
+	// Put all the new values in the DayStores under target
 	for key, tpSlice := range newMetrics {
 		_, ok := target.Metrics[key]
 		if !ok {
 			// Metric does not exist on target InfoType, create TimeStore
-			newTS := rc.tsConstructor()
-			target.Metrics[key] = &newTS
+			// TODO(afein): configure epsilon
+			newDS := daystore.NewDayStore(100, rc.resolution)
```
Comment (vishh) [inline model/aggregation.go]: Do you have a follow up PR for epsilon? `100` defeats the whole effort.

Comment (afein) [inline model/aggregation.go]: please do not review this PR yet, there are 650 failing assertions and a
set of refactoring changes that need to be put in first

File: model/types.go [hunk 4]
```diff
@@ -19,43 +19,52 @@ import (
...
 )
 
 type Cluster interface {
 	// The Update operation populates the Cluster from a cache.
 	Update(cache.Cache) error
 
+	// The simple Get operations extract structural information from the Cluster.
+	GetAvailableMetrics() []string
+	GetNodes() []EntityListEntry
+	GetNamespaces() []EntityListEntry
```
Comment (vishh) [inline model/types.go]: Ideally, only nodes and namespaces must be accessible from cluster. Those objects should then expose the pods that are under them. I am ok with this flat interface for now, but add a TODO.

**Comment by vishh** [comment]
Why add the new cadvisor deps?

File: model/aggregation.go (hunk 3 above)
Comment (vishh) [inline model/aggregation.go]: ok. Ping me once you have updated.

File: model/types.go (hunk 4 above)
Comment (afein) [inline model/types.go]: the interface can be renamed to Model to reflect its holistic scope of the cluster's structure. Having sub-objects would require a redesign of the current locking mechanisms, as we have discussed in the past

File: api/v1/model_handlers.go (hunk 1 above)
Comment (afein) [inline api/v1/model_handlers.go]: this should have been a pointer on the manager. Thanks for catching this!

**Comment by afein** [comment]
An update on this PR:

At the moment, I'm trying to bring this to a condition where tests are passing.
After that, most of the points that @vishh has mentioned will be also addressed, along with other leftover issues.

The majority of this PR includes Model10, which was submitted independently 4-5 weeks ago.
Now that the underlying stores have changed, a large set of past tests are failing, mostly due to assumptions taken on the behavior of Get and Put with the old stores.

I invited an initial review on this PR mostly for structural comments on the implementation of derived stats. However, there are still a few things that are leftover:
- make tests passing
- compare accuracy/memory footprint with that of the the older stores
- cleanup of CAdvisor dependencies
- breakdown of this PR into readable commits.
- definition of epsilon per metric (initially hardcoded in types.go, with a TODO for a configuration object argument on the model constructor)

I will ping @vishh and @mvdan when these tasks are completed and I will remove the (WIP) prefix from this PR then. At that point, please review the PR commit-by-commit for readability and reduced mental contexts.
ETA: Tuesday 08/25

**Comment by jszczepkowski** [comment]
CC @jszczepkowski

**Comment by afein** [comment]
Quick Update:
This PR is almost fully operational. I've had to change the behavior of a few components to comply with the new stores.
As soon as this PR is mergeable, I will start submitting separate smaller PRs that link to this one to facilitate the code review process.

The changes that will be included in this set of PRs are:
- usage of the new StatStore and DayStore
- static epsilon configuration
- Uptime calculation bugfix
- Unit test refactoring for compliance
- renaming of the top-level model interface to "Model" instead of "Cluster" to avoid confusion with the ClusterInfo struct

Furthermore, based on preliminary testing, the memory usage of the Heapster Pod is now consistently between 30-40MB for a not-too-noisy cluster. Using the previous stores, the usage was 200MB instead, leading to an improvement of 5x (rough estimate).

**Comment by vishh** [comment]
Thanks for the update Alex! Looking forward to the PRs!

**Comment by afein** [comment]
retest this please

**Comment by afein** [comment]
@mvdan these are the travis and jenkins errors I was getting after my initial godep cleanup, which is why I closed that original PR.

I am able to build and test locally, so I'm guessing there is some mismatch between what the expected testing environment is and what it should be

**Comment by mvdan** [comment]
Makes sense, it's a dependency of fleet, which isn't removed by this cleanup. Then the question should be, why is it the local build working fine?

**Comment by afein** [comment]
fleet is being used in sources/nodes/ and clusters/, therefore it is reasonable that the build should fail.
The reason that the local build was passing is that the googleapi dependency was located in my gopath.

The main concern around this issue is the following:
- During the `godep save ./...` operation, the googleapi dependency is not included in Godeps, even though it is dependent by fleet

I will run a secondary check on a clean gopath to see if reinstating the googleapi dependency will resolve this issue alone

**Comment by mvdan** [comment]
Sounds like a bug in godep. `godep build` et al falling back to `GOPATH` is also very confusing and sounds like a misfeature to me.

**Comment by mvdan** [comment]
Hang on, are you running the latest godep? I already did the cleanup here: https://github.com/kubernetes/heapster/commit/8523d38ab030e9cc65864d7b0638f12d6a36639d

It cleaned up some libraries, but not that one. So your godep is doing extra cleanups compared to mine, which are wrong.

**Comment by afein** [comment]
You're right, seems to be a bug on my Godep version. I checked out all the Godeps from master and it seems to be working

File: deploy/kube-config/kubedash/heapster-controller.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	"template": {
+	    "metadata": {
+		"labels": {
+		    "name": "heapster"
+		}
+	    },
+	    "spec": {
+		"containers": [
+		    {
+			"image": "afein/heapster-model:latest",
```
Comment (piosz) [inline deploy/kube-config/kubedash/heapster-controller.json]: How about making this image more official?

Comment (afein) [inline deploy/kube-config/kubedash/heapster-controller.json]: That's a fair point. I no longer have access to a gcr.io bucket so someone else would have to actually deploy this.

These deployment files can be eventually deprecated, as soon as the next milestone is deployed in all kubernetes clusters. In that case, the model should be accessible by default in the heapster pod running in the kube-system namespace.

File: api/v1/model_handlers.go [hunk 6]
```diff
@@ -332,88 +395,124 @@ func (a *Api) podPaths(request *restful.Request, response *restful.Response) {
...
 		"metrics/",
+		"stats/",
 	}
 	response.WriteEntity(entities)
 }
 
-// allNodes returns a list of all the available node names in the cluster.
+// makeExternalEntityList converts the result of model's getX methods to the external type.
+func makeExternalEntityList(list []model_api.EntityListEntry) []types.ExternalEntityListEntry {
+	res := make([]types.ExternalEntityListEntry, 0)
```
Comment (mvdan) [inline api/v1/model_handlers.go]: `0, len(list)`

File: api/v1/types/model_types.go [hunk 7]
```diff
@@ -31,3 +31,27 @@ type MetricResult struct {
 type MetricResultList struct {
 	Items []MetricResult `json:"items"`
 }
+
+type Stats struct {
+	Average    uint64 `json:"average"`
+	Percentile uint64 `json:"percentile"`
```
Comment (mvdan) [inline api/v1/types/model_types.go]: What percentile?

File: api/v1/types/model_types.go [hunk 8]
```diff
@@ -31,3 +31,27 @@ type MetricResult struct {
...
+
+type ExternalStatBundle struct {
+	Minute Stats `json:"minute"`
+	Hour   Stats `json:"hour"`
+	Day    Stats `json:"day"`
+}
+
+type StatsResponse struct {
+	// Uptime is in seconds
+	Uptime uint64                        `json:"uptime"`
```
Comment (mvdan) [inline api/v1/types/model_types.go]: Why not use time.Time or util.Time?

Comment (afein) [inline api/v1/types/model_types.go]: time.Time represents a time format, however the Uptime is a time.Duration field.
Since there is no standardized format for durations (unlike RFC3339 for time.Time), I opted to output the total number of seconds for that duration

File: api/v1/types/model_types.go (hunk 7 above)
Comment (afein) [inline api/v1/types/model_types.go]: 95th. I'll change that name, as well as the Stats struct in model/types.go
However, I will not change the json representation as of now to avoid conflicts in Kubedash

File: api/v1/model_handlers.go (hunk 6 above)
Comment (afein) [inline api/v1/model_handlers.go]: done

File: api/v1/types/model_types.go (hunk 8 above)
Comment (mvdan) [inline api/v1/types/model_types.go]: I was thinking of Duration, yes :) Sounds fair.

File: model/getters.go [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+	res := (*ts).Hour.Get(req.Start, req.End)
+	return res, rc.timestamp, nil
+}
+
+// makeEntityList creates an EntityListEntry from a map of metrics.
+func makeEntityListEntry(name string, entities map[string]*daystore.DayStore) EntityListEntry {
+	newListEntry := EntityListEntry{}
+	cpu, ok := entities[cpuUsage]
+	if !ok {
+		newListEntry.CPUUsage = uint64(0)
```
Comment (mvdan) [inline model/getters.go]: All the `uint64(0)` assignments are unnecessary since that's the zero value.

File: api/v1/model_handlers.go [hunk 10]
```diff
@@ -57,6 +57,13 @@ func (a *Api) RegisterModel(container *restful.Container) {
 		Doc("Get a list of all available metrics for the Cluster entity").
 		Operation("availableMetrics"))
 
+	// The /stats/ endpoint returns a list of all available stats for the Cluster entity of the model.
+	ws.Route(ws.GET("/stats/").
+		To(a.clusterStats).
+		Filter(compressionFilter).
+		Doc("Get all available stats for the Cluster entity").
+		Operation("clusterStats"))
```
Comment (mvdan) [inline api/v1/model_handlers.go]: Aren't all of these new endpoints missing `Writes()`?

Also, many other endpoints say `Writes(types.MetricResult{})` but that's just wrong, e.g. `GET("/nodes/")`

**Comment by afein** [comment]
Closing this PR. The feature set that was included has been broken down into #508 and #525
